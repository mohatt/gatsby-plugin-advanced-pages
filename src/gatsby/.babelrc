{
  "presets": [
    ["babel-preset-gatsby-package", { "nodeVersion": "8" }]
  ]
}

