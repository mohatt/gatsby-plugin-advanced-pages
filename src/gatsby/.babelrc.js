module.exports = {
  presets: [
    ['babel-preset-gatsby-package', {
      browser: false,
      nodeVersion: '18.0.0',
      esm: !!process.env.BUILD_ESM,
    }]
  ]
}
