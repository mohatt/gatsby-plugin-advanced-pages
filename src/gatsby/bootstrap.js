import { initializeOptions, initializeReporter } from './util'
import { optionsSchema } from './schema'

// Validates user-defined options against schema
// runs before onPreBootstrap
export function pluginOptionsSchema ({ Joi }) {
  return optionsSchema(Joi)
}

// Initializes plugin reporter
export function onPreInit ({ reporter }) {
  initializeReporter(reporter)
}

// Initializes plugin options
export function onPreBootstrap ({ store, reporter }, pluginOptions) {
  // Default values for options
  const defaultOptions = {
    basePath: '/',
    pages: [],
    template: null,
    directories: {
      templates: './src/templates',
      helpers: './gatsby/pages'
    },
    pagination: {
      limit: 10,
      suffix: '/page/:page'
    },
    typeNames: {
      page: 'Page'
    }
  }

  // Initializes and validates options
  // Only runs once at bootstrap
  initializeOptions({
    store,
    defaultOptions,
    pluginOptions
  })
}
