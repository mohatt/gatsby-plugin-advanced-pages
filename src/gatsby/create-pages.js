import path from 'path'
import PagesCreator from './lib/pages-creator'
import { getOption, reportError } from './util'

export default async function ({ graphql, actions, getNodesByType }) {
  const { createPage } = actions

  const pageType = getOption('typeNames.page')
  if(getNodesByType(pageType).length === 0) {
    return
  }

  const result = await graphql(`
    {
      all${pageType} {
        nodes {
          id
          routes {
            name
            path
          }
          title
          templateName
          template
          helper
        }
      }
    }
  `)

  if (result.errors) {
    return reportError('Failed running "create-pages" GraphQL Query', result.errors.shift())
  }

  const pageCreator = new PagesCreator(
    result.data[`all${pageType}`].nodes
  )

  // Create the actual pages
  await pageCreator.createPages({ graphql, createPage })

  // Write routes export file
  pageCreator.writeRoutesExport(path.resolve(__dirname, '../routes.js'))
}
