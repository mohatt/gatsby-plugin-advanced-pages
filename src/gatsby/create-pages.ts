import fs from 'fs'
import path from 'path'
import type { GatsbyNode } from 'gatsby'
import PagesCreator from './lib/pages-creator'
import { options, reporter } from './util'

const createPages: GatsbyNode['createPages'] = async ({ graphql, actions, getNodesByType, cache }) => {
  const { createPage } = actions

  const pageType = options.get('typeNames').page
  if (getNodesByType(pageType).length === 0) {
    return
  }

  const result = await graphql(`
    {
      all${pageType} {
        nodes {
          id
          routes {
            name
            path
          }
          title
          templateName
          template
          helper
        }
      }
    }
  `)

  if (result.errors) {
    reporter.error('Failed running "create-pages" GraphQL Query', result.errors.shift())
    return
  }

  // Create the actual pages
  const pageCreator = new PagesCreator(result.data[`all${pageType}`].nodes)
  await pageCreator.createPages({ graphql, createPage })

  // Write routes export file
  const routesFile = path.join(cache.directory, 'routes.json')
  try {
    fs.writeFileSync(routesFile, JSON.stringify(pageCreator.getRoutesExport()))
  } catch (e) {
    reporter.error('Error writing route map export file', e)
  }
}

export default createPages
