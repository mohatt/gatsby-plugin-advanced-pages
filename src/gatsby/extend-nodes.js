const { getOption } = require('./util')

// Extend frontmatter type definition
module.exports = ({ type }) => {
  if (type.name === getOption('engine.typeName')) {
    return {
      'frontmatter.title': 'String',
      'frontmatter.routes': 'JSON',
      'frontmatter.template': 'String',
      'frontmatter.helper': 'String',
      'frontmatter.data': 'JSON',
    }
  }

  return {}
}
