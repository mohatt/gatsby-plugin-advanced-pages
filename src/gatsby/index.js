export { default as onPreBootstrap } from './bootstrap'
export { default as sourceNodes } from './source-nodes'
export { default as setFieldsOnGraphQLNodeType } from './extend-nodes'
export { default as onCreateNode } from './on-create-node'
export { default as createPages } from './create-pages'
export { default as onPreExtractQueries } from './install-fragments'
