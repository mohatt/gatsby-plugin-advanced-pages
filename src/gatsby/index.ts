export type * from './types'
export type { RouteParams, PathGenerator } from '../lib/route-compiler'
