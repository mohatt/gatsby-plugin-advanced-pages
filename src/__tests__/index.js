import { withPrefix } from 'gatsby'
import mockRoutes from './__fixtures__/routes'
import router from '@reach/router'
import {
  getRoutes,
  routeExists,
  getRoute,
  getActivatedRoute,
  isActivatedRoute,
  getMatchingRoute,
  generatePath
} from '../index'

// Create a virtual mock for routes.json
jest.mock(
  'gatsby-plugin-advanced-pages-cache/routes.json',
  () => mockRoutes,
  { virtual: true }
)

router.useLocation = jest.fn().mockReturnValue({
  pathname: withPrefix('/blog/page/5')
})

describe('API', () => {
  it('correctly fetchs routes', () => {
    expect(getRoutes()).toMatchSnapshot()
    expect(getRoutes('blog')).toMatchSnapshot()
    expect(routeExists('blog.post')).toBe(true)
    expect(routeExists('invalid')).toBe(false)
    expect(getRoute('blog.post')).toMatchSnapshot()
    expect(() => getRoute('invalid')).toThrow()
    expect(() => getRoute(false)).toThrow()
    expect(() => getRoute(23)).toThrow()
  })

  it('picks the correct matching route', () => {
    expect(getActivatedRoute()).toMatchSnapshot()
    expect(isActivatedRoute('home')).toBe(false)
    expect(isActivatedRoute('blog')).toBe(true)
    expect(() => isActivatedRoute('invalid')).toThrow()
    expect(getMatchingRoute('/')).toMatchSnapshot()
    expect(getMatchingRoute(withPrefix('/'), true)).toMatchSnapshot()
    expect(getMatchingRoute('/blog/tag/test/page/4')).toMatchSnapshot()
    expect(getMatchingRoute('/blog/post/hi')).toMatchSnapshot()
  })

  it('correctly generates paths', () => {
    expect(generatePath('home')).toMatchSnapshot()
    expect(generatePath('blog.post', { post: 'hello' })).toMatchSnapshot()
    expect(() => generatePath('invalid')).toThrow()
  })
})
