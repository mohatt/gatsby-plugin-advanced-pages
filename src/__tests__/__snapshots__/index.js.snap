// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API correctly fetchs routes 1`] = `
Array [
  Object {
    "name": "home",
    "parent": null,
    "path": "/",
    "realpath": "/site/",
  },
  Object {
    "name": "about",
    "parent": null,
    "path": "/about",
    "realpath": "/site/about",
  },
  Object {
    "name": "blog",
    "parent": null,
    "path": "/blog",
    "realpath": "/site/blog",
  },
  Object {
    "name": "blog.pagination",
    "parent": Object {
      "name": "blog",
      "scope": "pagination",
    },
    "path": "/blog/page/:page",
    "realpath": "/site/blog/page/:page",
  },
  Object {
    "name": "blog.author",
    "parent": null,
    "path": "/blog/author/:author",
    "realpath": "/site/blog/author/:author",
  },
  Object {
    "name": "blog.author.pagination",
    "parent": Object {
      "name": "blog.author",
      "scope": "pagination",
    },
    "path": "/blog/author/:author/page/:page",
    "realpath": "/site/blog/author/:author/page/:page",
  },
  Object {
    "name": "blog.tag",
    "parent": null,
    "path": "/blog/tag/:tag",
    "realpath": "/site/blog/tag/:tag",
  },
  Object {
    "name": "blog.tag.pagination",
    "parent": Object {
      "name": "blog.tag",
      "scope": "pagination",
    },
    "path": "/blog/tag/:tag/page/:page",
    "realpath": "/site/blog/tag/:tag/page/:page",
  },
  Object {
    "name": "blog.post",
    "parent": null,
    "path": "/blog/post/:post",
    "realpath": "/site/blog/post/:post",
  },
]
`;

exports[`API correctly fetchs routes 2`] = `
Array [
  Object {
    "name": "blog.pagination",
    "parent": Object {
      "name": "blog",
      "scope": "pagination",
    },
    "path": "/blog/page/:page",
    "realpath": "/site/blog/page/:page",
  },
]
`;

exports[`API correctly fetchs routes 3`] = `
Object {
  "name": "blog.post",
  "parent": null,
  "path": "/blog/post/:post",
  "realpath": "/site/blog/post/:post",
}
`;

exports[`API correctly generates paths 1`] = `"/site"`;

exports[`API correctly generates paths 2`] = `"/site/blog/post/hello"`;

exports[`API picks the correct matching route 1`] = `
Object {
  "name": "blog.pagination",
  "parent": Object {
    "name": "blog",
    "scope": "pagination",
  },
  "path": "/blog/page/:page",
  "realpath": "/site/blog/page/:page",
}
`;

exports[`API picks the correct matching route 2`] = `
Object {
  "name": "home",
  "parent": null,
  "path": "/",
  "realpath": "/site/",
}
`;

exports[`API picks the correct matching route 3`] = `
Object {
  "name": "home",
  "parent": null,
  "path": "/",
  "realpath": "/site/",
}
`;

exports[`API picks the correct matching route 4`] = `
Object {
  "name": "blog.tag.pagination",
  "parent": Object {
    "name": "blog.tag",
    "scope": "pagination",
  },
  "path": "/blog/tag/:tag/page/:page",
  "realpath": "/site/blog/tag/:tag/page/:page",
}
`;

exports[`API picks the correct matching route 5`] = `
Object {
  "name": "blog.post",
  "parent": null,
  "path": "/blog/post/:post",
  "realpath": "/site/blog/post/:post",
}
`;
